crates/ruff_python_semantic/src/cfg/graph.rs

```diff
@@ -0,0 +1,282 @@
+use ruff_index::{newtype_index, IndexVec};
+use ruff_python_ast::Stmt;
+use ruff_text_size::{Ranged, TextRange};
+use smallvec::{smallvec, SmallVec};
+
+/// Returns the control flow graph associated to an array of statements
+pub fn build_cfg(stmts: &[Stmt]) -> CFG<'_> {
+    let mut builder = CFGBuilder::with_capacity(stmts.len());
+    builder.process_stmts(stmts);
+    builder.finish()
+}
+
+/// Control flow graph
+#[derive(Debug)]
+pub struct CFG<'stmt> {
+    /// Basic blocks - the nodes of the control flow graph
+    blocks: IndexVec<BlockId, BlockData<'stmt>>,
+    /// Entry point to the control flow graph
+    initial: BlockId,
+    /// Terminal node - will always be empty
+    terminal: BlockId,
+}
+
+impl<'stmt> CFG<'stmt> {
+    /// Index of entry point to the control flow graph
+    pub fn initial(&self) -> BlockId {
+        self.initial
+    }
+
+    /// Index of terminal node
+    pub fn terminal(&self) -> BlockId {
+        self.terminal
+    }
+
+    /// Number of basic blocks, or nodes, in the graph
+    pub fn num_blocks(&self) -> usize {
+        self.blocks.len()
+    }
+
+    /// Returns the statements comprising the basic block at the given index
+    pub fn stmts(&self, block: BlockId) -> &'stmt [Stmt] {
+        self.blocks[block].stmts
+    }
+
+    /// Returns the range of the statements comprising the basic block at the given index
+    pub fn range(&self, block: BlockId) -> TextRange {
+        self.blocks[block].range()
+    }
+
+    /// Returns the [`Edges`] going out of the basic block at the given index
+    pub fn outgoing(&self, block: BlockId) -> &Edges {
+        &self.blocks[block].out
+    }
+
+    /// Returns an iterator over the indices of the direct predecessors of the block at the given index
+    pub fn predecessors(&self, block: BlockId) -> impl ExactSizeIterator<Item = BlockId> + '_ {
+        self.blocks[block].parents.iter().copied()
+    }
+
+    /// Returns the [`BlockKind`] of the block at the given index
+    pub(crate) fn kind(&self, block: BlockId) -> BlockKind {
+        self.blocks[block].kind
+    }
+}
+
+#[newtype_index]
+pub struct BlockId;
+
+/// Holds the data of a basic block. A basic block consists of a collection of
+/// [`Stmt`]s, together with outgoing edges to other basic blocks.
+#[derive(Debug, Default)]
+struct BlockData<'stmt> {
+    kind: BlockKind,
+    stmts: &'stmt [Stmt],
+    out: Edges,
+    parents: SmallVec<[BlockId; 2]>,
```

Comment:
Can you add some documentation what `parents` are in a basic block context? Is it the incoming basic blocks?